src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs

```diff
@@ -808,9 +808,12 @@ public TProperty GetRelationshipSnapshotValue<TProperty>(IPropertyBase propertyB
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
         public object? ReadPropertyValue(IPropertyBase propertyBase)
-            => propertyBase.IsShadowProperty()
-                ? _shadowValues[propertyBase.GetShadowIndex()]
-                : propertyBase.GetGetter().GetClrValue(Entity);
+            => propertyBase.IsPseudoProperty
+                ? ((IPseudoProperty)propertyBase).ValueExtractor(
+                    ReadPropertyValue(((IPseudoProperty)propertyBase).OuterProperty))
```

Comment:
Why not repurpose the Getter?
```suggestion
                ? propertyBase.GetGetter().GetClrValue(ReadPropertyValue(((IPseudoProperty)propertyBase).OuterProperty))
```

Reply:
How will the shadow pseudo properties handled?

Reply:
I don't think shadow pseudo properties are a thing. A shadow property has storage in the state manager and no storage in the CLR type. In this case, the property has no storage of it's own because the value doesn't exist in this form in application code at all.

Reply:
Yes, the outer property is basically a complex property if the CLR type contains properties that can be used for pseudo-properties. If the outer type is something like a string then the pseudo-properties don't map to a CLR property but still only keep state in the entity object. We might need a different name for this.